Bugfixes and testing of linear templates

jsiirola opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE. Failing to adhere to this template and provide the necessary information may lead to your Pull Request being closed without consideration. -->

## Fixes # .

## Summary/Motivation:
This PR is primarily adding  more comprehensive testing of templatizing (and compiling) linear models.  As part of flusing out the tests, this resolves a number of issues / bugs in the templatizer, the linear template compiler, and in general Pyomo:

General Pyomo changes
- Rework `Param.extract_values` to return `defaultdict` for Params with default values
- Provide a specialized `InvalidConstraintError` exception for constraints that cannot be emitted to the solvers
- Use `SetOf` when iterating over IndexedComponent objects indexed by non-finite Sets
- `SetOf.dimen` should return `UnknownSetDimen` (*not* 0) if the underlying reference has no data

Updates to the templatizer
- Leverage `attempt_import` to break circular dependencies in `template_expr` 
- Catch `UnknownSetDimen` when iterating over template generators and treat like jagged sets
- Expand (do not templatize) expressions with `sum()` that do not sum over generators
- Expand (do not templatize) expressions with `sum()` whose generator doesn't create a template index
- Expand (do not templatize) expressions with `sum()` that appear to iterate over things other than Pyomo Set objects

Updates to the linear template compiler
- remove `check_duplicates` and `remove_fixed_vars` options (this is not more efficiently implementable through sparse matrix operations0
- rework expression builders to be more testable and (slightly) more efficient
- Support `IndexedComponent.Skip`
- Resolve issue compiling constraints indexed by non-finite Sets

## Changes proposed in this PR:
- (see above)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

jsiirola committed:
Define InvalidConstraintError for unallowed constraint expressions

jsiirola committed:
Remove check_duplicates from linear template compiler

jsiirola committed:
When compiling templates, call to_string in everything that is not a native type

jsiirola committed:
Catch (and test) bug where TemplateVarRecorder ignored the sorter

jsiirola committed:
Avoid awkward call to base class by just reimplementing the method

jsiirola committed:
Use attempt_import to break circular dependencies

jsiirola committed:
Start tests for repn/linear_template

jsiirola committed:
bugfix: converting LinearTemplateRepn to str

jsiirola committed:
Improve specificity of raised exceptions

jsiirola committed:
Fix filter for 0 multipliers/coefs

jsiirola committed:
Correctly handle empty subexpressions (e.g., from 0 multipliers)

jsiirola committed:
Explicitly disallow nonlinear expressions in the LinearTemplateRepn

jsiirola committed:
Use SetOf when iterating over non-Set indexing sets

jsiirola committed:
Rework extract_values to use defaultdict for Params with default values

jsiirola committed:
SetOf.dimen updates:

- iterate over the underlying reference when guessing the dimen
  This prevents recursion when templatizing.
- inconsistent dimen is reported as None

jsiirola committed:
Ensure we only attempt to templatize generators

jsiirola committed:
Expand summations that do not involve IndexTemplates

jsiirola committed:
Improve error reporting

jsiirola committed:
Rework linear_template to enable better unit testing

jsiirola committed:
Resolve handling of fixed variables in bounds and skipped constraints

jsiirola committed:
Don't unnecessarily expand sparse Params

jsiirola committed:
Improve error

jsiirola committed:
Verify absence of bounds for objectives; update comment

jsiirola committed:
Add interface to get var_list for TemplateVarRecorder

jsiirola committed:
Expand linear_template testing

jsiirola committed:
Merge branch 'main' into linear-template-testing

jsiirola committed:
SetOf: revert change to the dimen of empty iterables

jsiirola committed:
Update tests to reflect use of SetOf; remove deprecation warnings

jsiirola committed:
Revert "SetOf: revert change to the dimen of empty iterables"

This reverts commit c7c5d6f63313f23bb17473832f2bc5c2c2b5d19f.

jsiirola committed:
Track change that SetOf([]).dimen is now None

jsiirola committed:
SetOf({}).dimen should return UnknownSetDimen

jsiirola committed:
Merge branch 'main' into linear-template-testing

jsiirola committed:
Avoid slicing over UnknownSetDimen referenecs

jsiirola committed:
Explicit sum() should not prevent templatizing constraint

jsiirola committed:
Additional validation before creating TemplateSumExpression objects

jsiirola committed:
Remove (unused) 'remove_fixed_vars' option

jsiirola committed:
Remove duplicated code

jsiirola committed:
NFC: fix typo

jsiirola committed:
Merge branch 'main' into linear-template-testing

emma58 commented on pyomo/core/base/constraint.py:
```diff
@@ -219,7 +223,7 @@ def to_bounded_expression(self, evaluate_bounds=False):
+                raise InvalidConstraintError(
                     f"Constraint '{self.name}' is a Ranged Inequality with a "
                     "variable lower bound.  Cannot normalize the "
                     "constraint or send it to a solver."
```
While we're changing things here, is 'normalize' the right word? Maybe the second sentence of this error should be much simpler: "Pyomo doesn't allow Constraints of the form lb <= body <= ub where either of lb or ub contains variables."

jsiirola committed:
Merge branch 'main' into linear-template-testing

jsiirola committed:
NFC: Apply black

jsiirola committed:
Catch errors iterating over component values in templates

emma58 commented on pyomo/repn/tests/test_linear_template.py:
```diff
@@ -0,0 +1,1227 @@
+from pyomo.repn.linear_template import LinearTemplateRepnVisitor
+from pyomo.repn.util import TemplateVarRecorder
+
+from pyomo.environ import *
```
Miranda's going to be mad at you!

jsiirola committed:
Merge branch 'main' into linear-template-testing

emma58 reviewed (commented):
I kind of ran out of brainpower in the crux, so I'll look at `linear_template.py` again in the morning, but here are a few comments from elsewhere. I'm very exited that it's tested!!

jsiirola committed:
Address review comments

emma58 reviewed (approved):
OK, please don't get hit by a bus, but I kind of understand this, and I reallllly like it!

jsiirola committed:
Merge branch 'main' into linear-template-testing

jsiirola committed:
Address review: rename normalize -> standardize

jsiirola committed:
Add InvalidExpressionError

This supports improved ValueError exceptions from code processing
general expressions (i.e., it could be an objective or constraint
expression)

jsiirola committed:
Raise an InvalidConstraintError and not RuntimeError for non-fixed bounds

jsiirola committed:
Log an informative error for invalid linear template expressions

jsiirola committed:
NFC: doc update

mrmundt commented on pyomo/repn/tests/test_linear_template.py:
```diff
@@ -0,0 +1,1257 @@
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
```
Oh noooo, John... Old copyright. I'll go fix it, this comment is just a heads up

mrmundt committed:
Outdated copyright statement

jsiirola committed:
Merge branch 'main' into linear-template-testing

jsiirola committed:
Merge branch 'main' into linear-template-testing

blnicho commented on pyomo/repn/linear_template.py:
```diff
@@ -37,14 +46,29 @@ def __init__(self):
             f"linear={linear}, "
-            f"linear_sum={self.linear_sum}, "
+            f"linear_sum={linear_sum}, "
             f"nonlinear={self.nonlinear})"
```
Do you need the nonlinear line in the string representation if we're asserting that `self.nonlinear is None` for linear templates?

blnicho reviewed (approved):
I have a question but otherwise this can be merged!
